### PR TITLE
ionoscloud: Update ionos-cloud-sdk-go and add metrics

### PR DESCRIPTION
**Pull request by avorima**
#### What type of PR is this?

/kind cleanup
/kind feature

#### What this PR does / why we need it:

Address a few things that built up over time:
* Update ionos-cloud/sdk-go to v6.1.11
* Fix example values link in provider README
* Configurable additional request headers
* Add cloud API response metrics

```
# HELP cluster_autoscaler_ionoscloud_api_request_total [ALPHA] Counter of IonosCloud API requests for each action and response status.
# TYPE cluster_autoscaler_ionoscloud_api_request_total counter
cluster_autoscaler_ionoscloud_api_request_total{action="DeleteNode",status="success"} 1
cluster_autoscaler_ionoscloud_api_request_total{action="GetNodePool",status="success"} 12
cluster_autoscaler_ionoscloud_api_request_total{action="ListNodes",status="success"} 10
cluster_autoscaler_ionoscloud_api_request_total{action="ResizeNodePool",status="success"} 1
```

#### Does this PR introduce a user-facing change?

```release-note
Add `cluster_autoscaler_ionoscloud_api_request_total` alpha metric
```

### CODE REVIEW

File: cluster-autoscaler/cloudprovider/ionoscloud/mock_api_client_test.go [hunk 1]
```diff
@@ -14,7 +14,7 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-// Code generated by mockery 2.12.1. DO NOT EDIT.
+// Code generated by mockery v2.24.0. DO NOT EDIT.
```
Comment (piepmatz) [inline cluster-autoscaler/cloudprovider/ionoscloud/mock_api_client_test.go]: Looks like an upgrade from stone age to iron age. :P Any reason for that specific version?

File: cluster-autoscaler/cloudprovider/ionoscloud/cache.go [hunk 2]
```diff
@@ -94,23 +76,30 @@ func (cache *IonosCache) SetInstancesCacheForNodeGroup(id string, instances []cl
 	cache.setInstancesCacheForNodeGroupNoLock(id, instances)
 }
 
+func (cache *IonosCache) removeNodesForNodeGroupNoLock(id string) {
+	for nodeID, nodeGroupID := range cache.nodesToNodeGroups {
+		if nodeGroupID == id {
+			delete(cache.nodesToNodeGroups, nodeID)
+		}
+	}
```
Comment (piepmatz) [inline cluster-autoscaler/cloudprovider/ionoscloud/cache.go]: https://pkg.go.dev/maps@go1.21.8#DeleteFunc maybe?

File: cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager_test.go [hunk 3]
```diff
@@ -283,45 +316,45 @@ func (s *ManagerTestSuite) TestSetNodeGroupSize_ResizeError() {
...
 			pollCount++
 		})
 
 	err := s.manager.SetNodeGroupSize(s.nodePool, 2)
 	s.Error(err)
-	s.True(errors.Is(err, wait.ErrWaitTimeout))
+	s.True(wait.Interrupted(err))
 	// The poll count may vary, so just do this to prevent flakes.
-	s.True(pollCount > int(pollTimeout/pollInterval))
+	s.GreaterOrEqual(pollCount, int(pollTimeout/pollInterval))
```
Comment (piepmatz) [inline cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager_test.go:340]: You changed from `>` to `≥`. Intentionally?

File: cluster-autoscaler/cloudprovider/ionoscloud/cache.go [hunk 4]
```diff
@@ -18,28 +18,18 @@ package ionoscloud
...
 	"k8s.io/autoscaler/cluster-autoscaler/cloudprovider"
 	"k8s.io/klog/v2"
 )
 
-const nodeGroupCacheEntryTimeout = 2 * time.Minute
-
-type nodeGroupCacheEntry struct {
-	data cloudprovider.NodeGroup
-	ts   time.Time
-}
```
Comment (piepmatz) [inline cluster-autoscaler/cloudprovider/ionoscloud/cache.go:32]: Pls gimme some background on this. This looks to me as if we no longer require to refresh any cached node pools after 2m. Why?

File: cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager_test.go [hunk 5]
```diff
@@ -85,22 +91,24 @@ func TestLoadConfigFromEnv(t *testing.T) {
...
-				envKeyToken:        "token",
-				envKeyTokensPath:   "/etc/passwd",
+				envKeyClusterID:     "test",
+				envKeyEndpoint:      "/dev/null",
+				envKeyInsecure:      "1",
+				envKeyPollInterval:  "42ms",
+				envKeyPollTimeout:   "1337s",
+				envKeyToken:         "token",
+				envKeyTokensPath:    "/etc/passwd",
+				envKeyCustomHeaders: "foo:bar;baz:qux",
```
Comment (piepmatz) [inline cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager_test.go]: The required header format should be documented somewhere. Readme?

File: cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager.go [hunk 6]
```diff
@@ -276,18 +286,8 @@ func (manager *ionosCloudManagerImpl) GetInstancesForNodeGroup(nodeGroup cloudpr
 }
 
 func (manager *ionosCloudManagerImpl) GetNodeGroupForNode(node *apiv1.Node) cloudprovider.NodeGroup {
-	nodeId := convertToNodeId(node.Spec.ProviderID)
-	return manager.cache.GetNodeGroupForNode(nodeId)
-}
-
-// Refreshes the cache holding the instances for the configured node groups.
-func (manager *ionosCloudManagerImpl) Refresh() error {
```
Comment (piepmatz) [inline cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager.go:284]: Similar story here: What's the reasoning for removing this?

File: cluster-autoscaler/cloudprovider/ionoscloud/client.go [hunk 7]
```diff
@@ -65,87 +57,72 @@ type APIClient interface {
...
-		cloudConfigDir: config.TokensPath,
-		endpoint:       config.Endpoint,
-		userAgent:      userAgent,
-		insecure:       config.Insecure,
+func setLogLevel(config *ionos.Configuration) {
+	switch {
+	case klog.V(7).Enabled():
+		config.LogLevel = ionos.Trace
+	case klog.V(5).Enabled():
+		config.LogLevel = ionos.Debug
```
Comment (piepmatz) [inline cluster-autoscaler/cloudprovider/ionoscloud/client.go:89]: Is that something worth documenting?

File: cluster-autoscaler/cloudprovider/ionoscloud/client.go [hunk 8]
```diff
@@ -160,80 +137,87 @@ func (p customClientProvider) GetClient() (APIClient, error) {
 	if len(cloudConfig.Tokens) == 0 {
 		return nil, fmt.Errorf("missing tokens for cloud config %s", filepath.Base(files[0]))
 	}
-	return NewAPIClient(cloudConfig.Tokens[0], p.endpoint, p.userAgent, p.insecure), nil
+	cfg := *c.cfg
+	cfg.Token = cloudConfig.Tokens[0]
+	return NewAPIClient(&cfg, c.userAgent), nil
```
Comment (piepmatz) [inline cluster-autoscaler/cloudprovider/ionoscloud/client.go:142]: Looking at
```go
if c.client != nil {
	return c.client, nil
}
```
in the beginning, shouldn't we store this created client in `c.client`? Or don't we to support changing tokens without a restart?

File: cluster-autoscaler/cloudprovider/ionoscloud/client.go [hunk 9]
```diff
@@ -160,80 +137,87 @@ func (p customClientProvider) GetClient() (APIClient, error) {
...
-	client, err := c.GetClient()
+	client, err := c.getClient()
 	if err != nil {
 		return nil, err
 	}
-	req := client.K8sNodepoolsFindById(context.Background(), c.clusterId, id)
-	nodepool, _, err := client.K8sNodepoolsFindByIdExecute(req)
+	req := client.K8sNodepoolsFindById(context.Background(), c.cfg.ClusterID, id)
+	nodepool, resp, err := client.K8sNodepoolsFindByIdExecute(req)
+	registerResponse("nodepools", "get", resp)
```
Comment (piepmatz) [inline cluster-autoscaler/cloudprovider/ionoscloud/client.go]: Using consts here would feel a bit safer.

**Review by piepmatz — CHANGES_REQUESTED** [review]
The ticket mentions "Allow deploying in-cluster autoscaler in staging". Is that reflected in here somehow? If so, where?

**Comment by piepmatz** [comment]
The PR description's metrics example looks outdated.

File: cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager_test.go (hunk 3 above)
Comment (avorima) [reply cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager_test.go:340]: The test actually failed once with just Greater

File: cluster-autoscaler/cloudprovider/ionoscloud/cache.go (hunk 4 above)
Comment (avorima) [reply cluster-autoscaler/cloudprovider/ionoscloud/cache.go:32]: it's never used

File: cluster-autoscaler/cloudprovider/ionoscloud/mock_api_client_test.go (hunk 1 above)
Comment (avorima) [reply cluster-autoscaler/cloudprovider/ionoscloud/mock_api_client_test.go]: i'll find a version that still uses go 1.21

File: cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager_test.go (hunk 5 above)
Comment (avorima) [reply cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager_test.go]: I wouldn't know what example to give, so I left it out

File: cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager.go (hunk 6 above)
Comment (avorima) [reply cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager.go:284]: Also unused, but we might consider re-adding this at a later stage

File: cluster-autoscaler/cloudprovider/ionoscloud/client.go (hunk 8 above)
Comment (avorima) [reply cluster-autoscaler/cloudprovider/ionoscloud/client.go:142]: This if condition is intended for IONOS_TOKEN input, so not for our use case. This method of getting the client already allows changing tokens without a restart. It's just not as reliable as our later implementations. I wanted to keep things simple here

File: cluster-autoscaler/cloudprovider/ionoscloud/client.go (hunk 9 above)
Comment (avorima) [reply cluster-autoscaler/cloudprovider/ionoscloud/client.go]: Just followed the example from GCE/AWS

**Comment by avorima** [comment]
> The PR description's metrics example looks outdated.

Ah, seems I forgot to push my latest changes.

**Comment by avorima** [comment]
> The ticket mentions "Allow deploying in-cluster autoscaler in staging". Is that reflected in here somehow? If so, where?

The client is always constructed using config, which means that the endpoint is used even when IONOS_TOKEN is set.

File: cluster-autoscaler/cloudprovider/ionoscloud/README.md [hunk 10]
```diff
@@ -40,6 +40,11 @@ mockery --inpackage --testonly --case snake --boilerplate-file ../../../hack/boi
 mockery --inpackage --testonly --case snake --boilerplate-file ../../../hack/boilerplate/boilerplate.generatego.txt --name IonosCloudManager
 ```
 
+### Debugging
+
+The global logging verbosity controlled by the `--v` flag is passed on to the Ionos Cloud SDK client logger.
+Verbosity 5 maps to `Debug` and 7 to `Trace`. **Do not enable this in production as it will print full request and response data.**
```
Comment (piepmatz) [inline cluster-autoscaler/cloudprovider/ionoscloud/README.md]: place a comma before "as"

File: cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager_test.go (hunk 5 above)
Comment (piepmatz) [reply cluster-autoscaler/cloudprovider/ionoscloud/ionoscloud_manager_test.go]: How about `X-Contract-Number`? (https://api.ionos.com/docs/cloud/v6/)

File: cluster-autoscaler/cloudprovider/ionoscloud/client.go (hunk 8 above)
Comment (piepmatz) [reply cluster-autoscaler/cloudprovider/ionoscloud/client.go:142]: alright, thx

**Review by piepmatz — CHANGES_REQUESTED** [review]

**Comment by piepmatz** [comment]
LGTM

**Pull request closed without merging**